File: src/game/deck.ts
import { Card, RANKS, SUITS, Rank } from "./types";

export function createDeck(): Card[] {
  const deck: Card[] = [];
  for (const s of SUITS) {
    for (const r of RANKS) {
      deck.push({ id: `${r}${s}`, suit: s, rank: r });
    }
  }
  return deck;
}

export function shuffle<T>(arr: T[]): T[] {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

export const RANK_VALUE: Record<Rank, number> = {
  "2": 2,
  "3": 3,
  "4": 4,
  "5": 5,
  "6": 6,
  "7": 7,
  "8": 8,
  "9": 9,
  "10": 10,
  J: 11,
  Q: 12,
  K: 13,
  A: 14,
};

export function deal(numPlayers: number, cardsPerPlayer: number): Card[][] {
  const deck = shuffle(createDeck());
  const hands: Card[][] = Array.from({ length: numPlayers }, () => []);
  let idx = 0;
  for (let c = 0; c < cardsPerPlayer; c++) {
    for (let p = 0; p < numPlayers; p++) {
      hands[p].push(deck[idx++]);
    }
  }
  return hands;
}

export function sortHand(hand: Card[]): Card[] {
  const suitOrder: Record<string, number> = { S: 0, H: 1, C: 2, D: 3 };
  return [...hand].sort((a, b) => {
    if (a.suit !== b.suit) return suitOrder[a.suit] - suitOrder[b.suit];
    return RANK_VALUE[b.rank] - RANK_VALUE[a.rank];
  });
}

File: backend/app/games/kachuful/deck.py
"""Card deck operations — direct port of src/game/deck.ts."""
from __future__ import annotations

import random
from typing import Dict, List

from .types import Card, RANKS, Rank, SUITS


RANK_VALUE: Dict[Rank, int] = {
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "10": 10,
    "J": 11,
    "Q": 12,
    "K": 13,
    "A": 14,
}


def create_deck() -> List[Card]:
    return [Card(id=f"{r}{s}", suit=s, rank=r) for s in SUITS for r in RANKS]


def shuffle(cards: List[Card]) -> List[Card]:
    """Return a new shuffled copy — never mutate the input."""
    out = list(cards)
    random.shuffle(out)
    return out


def deal(num_players: int, cards_per_player: int) -> List[List[Card]]:
    """Deal one card at a time, round-robin, just like a human dealer."""
    deck = shuffle(create_deck())
    hands: List[List[Card]] = [[] for _ in range(num_players)]
    idx = 0
    for _ in range(cards_per_player):
        for p in range(num_players):
            hands[p].append(deck[idx])
            idx += 1
    return hands


def sort_hand(hand: List[Card]) -> List[Card]:
    """Match the frontend ordering: S, H, C, D — then by rank DESC inside each suit."""
    suit_order: Dict[str, int] = {"S": 0, "H": 1, "C": 2, "D": 3}
    return sorted(
        hand,
        key=lambda c: (suit_order[c.suit], -RANK_VALUE[c.rank]),
    )
